src/node/desktop/src/main/gwt-callback.ts

```diff
@@ -91,19 +91,46 @@ export class GwtCallback extends EventEmitter {
       }
     });
 
-    ipcMain.handle('desktop_get_save_file_name', (event, caption: string, label: string,
+    ipcMain.handle('desktop_get_save_file_name', async (event, caption: string, label: string,
       dir: string, defaultExtension: string, forceDefaultExtension: boolean,
       focusOwner: boolean
     ) => {
-      GwtCallback.unimpl('desktop_get_save_file_name');
-      return '';
+      const saveDialogOptions: SaveDialogOptions = {
+        title: caption,
+        defaultPath: dir,
+        buttonLabel: label
+      };
+
+      let focusedWindow = BrowserWindow.getFocusedWindow();
+      if (focusOwner) {
+        focusedWindow = this.getSender('desktop_open_minimal_window', event.processId, event.frameId).window;
+      }
+      if (focusedWindow) {
+        return dialog.showSaveDialog(focusedWindow, saveDialogOptions);
+      } else {
+        return dialog.showSaveDialog(saveDialogOptions);
+      }
     });
 
-    ipcMain.handle('desktop_get_existing_directory', (event, caption: string, label: string,
+    ipcMain.handle('desktop_get_existing_directory', async (event, caption: string, label: string,
       dir: string, focusOwner: boolean
     ) => {
-      GwtCallback.unimpl('desktop_get_existing_directory');
-      return '';
+      const openDialogOptions: OpenDialogOptions = {
+        title: caption,
+        defaultPath: dir,
+        buttonLabel: label,
+        properties: ['openDirectory', 'createDirectory', 'promptToCreate']
+      };
+
+      let focusedWindow = BrowserWindow.getFocusedWindow();
+      if (focusOwner) {
+        focusedWindow = this.getSender('desktop_open_minimal_window', event.processId, event.frameId).window;
+      }
+      if (focusedWindow) {
+        return dialog.showOpenDialog(focusedWindow, openDialogOptions);
+      } else {
+        return dialog.showOpenDialog(openDialogOptions);
+      }
```

Comment:
I just noticed that if we call `showOpenDialog` with `focusedWindow`, a dialog shows up in the middle of the main electron window, and it cannot be moved. The only way to "move" it is to move the main electron window, and the dialog window will move with the main window.

Personally I don't like it when sub-windows are locked in place, is this the behavior we want? I suggest removing most this logic and replacing it with:
`return dialog.showOpenDialog(openDialogOptions);`

Reply:
With that command, it creates the same dialog, but it's free to be moved around anywhere on the screen/other displays

Reply:
That is a Mac-specific thing, mentioned here: https://www.electronjs.org/docs/latest/api/dialog#sheets

We do need these dialogs to be modal, if you make them modeless (by not supplying the owning BrowserWindow), then you can bring up the dialog and continue interacting with the main window, which puts the app in an undefined state.

For example, you can bring up multiple File Open dialogs at once, and you can put the main window over a file-open dialog, both of which are Bad Things.

On Windows and Linux, the modal is not attached to its parent window like it is on the Mac, but the parent window is disabled while the dialog is displayed, and cannot be put under the parent window.

Unfortunately, I don't see a way to tell Electron to not make these modal dialogs be sheets on MacOS. In Qt there was a flag for it (`Qt::Sheet`). I didn't look very hard, though.